File: server-java/src/main/java/com/das/java/mock/MockFixtures.java
package com.das.java.mock;

import com.das.java.core.Frames;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * MOCK 模式的数据源：`server-node/test/fixtures/` 下的录制件（同一仓库、同一份数据，不复制第二份）。
 * 与 Node 实现的 server-node/mock/fixtures.ts、Python 实现的 mock_fixtures.py 同源同语义。
 */
public final class MockFixtures {
    private MockFixtures() {}

    private static final ObjectMapper MAPPER = new ObjectMapper();

    /** fixtures 在 Node server 的测试目录里。repoRoot 解析失败时这里拿不到文件，启动 MOCK 会报错——如实报。 */
    public static Path fixtureDir(Path repoRoot) {
        return repoRoot.resolve("server-node").resolve("test").resolve("fixtures");
    }

    public static List<Map<String, Object>> readFixtureFrames(Path repoRoot, String name) {
        Path file = fixtureDir(repoRoot).resolve(name);
        List<String> lines;
        try {
            lines = Files.readAllLines(file);
        } catch (IOException e) {
            throw new IllegalStateException("读 fixture 失败：" + file + "（" + e.getMessage() + "）", e);
        }
        List<Map<String, Object>> frames = new ArrayList<>();
        for (String line : lines) {
            if (line.trim().isEmpty()) continue;
            Map<String, Object> frame = Frames.parseRecordedLine(line);
            if (frame == null) {
                throw new IllegalStateException("fixture " + name + " 有一行解析不出帧");
            }
            frames.add(frame);
        }
        return frames;
    }

    /** 非流式录制件：形如 {JsonRpcResponse:{Result:{…}}, RequestId:'…'}。 */
    @SuppressWarnings("unchecked")
    public static Map<String, Object> readFixtureEnvelope(Path repoRoot, String name) {
        Path file = fixtureDir(repoRoot).resolve(name);
        Map<String, Object> parsed;
        try {
            parsed = MAPPER.readValue(Files.readString(file), Map.class);
        } catch (IOException e) {
            throw new IllegalStateException("读 fixture 失败：" + file + "（" + e.getMessage() + "）", e);
        }
        Map<String, Object> rpc = Frames.asMap(parsed.get("JsonRpcResponse"));
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("result", rpc == null ? null : rpc.get("Result"));
        out.put("requestId", parsed.get("RequestId"));
        out.put("error", rpc == null ? null : rpc.get("Error"));
        return out;
    }

    public static Map<String, Object> readFixtureResult(Path repoRoot, String name) {
        Map<String, Object> envelope = readFixtureEnvelope(repoRoot, name);
        Object result = envelope.get("result");
        if (!(result instanceof Map<?, ?> map)) {
            throw new IllegalStateException("fixture " + name + " 里没有 JsonRpcResponse.Result");
        }
        @SuppressWarnings("unchecked")
        Map<String, Object> typed = (Map<String, Object>) map;
        return typed;
    }

    public record MockScenario(
        String sessionId,
        String title,
        String promptFixture,
        String historyFixture,
        String popAck,
        int frameCount,
        String teaches,
        long createdAt,
        String sourceOverride
    ) {}

    public record CreateResult(String sessionId) {}

    private static final String SHORT = "mock-short";
    private static final String TOOLS = "mock-tools";
    private static final String LONG = "mock-long";
    private static final String BREAK = "mock-break";
    private static final String GHOST = "mock-ghost";
    private static final String CONCURRENT = "mock-concurrent";
    private static final String ACK = "mock-ack-only";
    private static final String RENDER = "mock-render";

    /** 八条演示会话：七条各对一份录制件（ack-only 那条没有，它演示的正是"零帧"）。 */
    public static final List<MockScenario> SCENARIOS = List.of(
        new MockScenario(SHORT, "[MOCK] 短轮 · 22 帧 · end_turn",
            "prompt-short.jsonl", "load-clean.jsonl", null, 22,
            "最短闭环：一问一答、无工具调用、5.6s 拿到 end_turn", 1_789_026_610_000L, null),
        new MockScenario(TOOLS,
            // marker 注入已退役：标题不再带校验码说明。
            "[MOCK] 工具轮 · 392 帧 · 6 次调用",
            "prompt-tools.jsonl", "load-clean.jsonl", null, 392,
            "工具状态机：6 次调用（5 completed + 1 failed）", 1_789_029_139_000L, null),
        new MockScenario(LONG, "[MOCK] 长轮 · 901 帧 · 191s · 10 次调用",
            "prompt-long.jsonl", "load-polluted.jsonl", null, 901,
            "长轮 191s / 901 帧；历史用 RUNNING 期录的那份（有 rid-less 污染）", 1_789_027_001_000L, null),
        new MockScenario(BREAK, "[MOCK] 断流 · -32603 · 1201 帧",
            "error-stream-break.jsonl", "load-polluted.jsonl", null, 1201,
            "SSE 断流：吐完 1200 帧后收尾 -32603，任务可能仍在服务端跑", 1_788_787_752_000L, null),
        new MockScenario(GHOST, "[MOCK] 会话幽灵化 · 422 · 单帧",
            "error-session-ghost.jsonl", null, null, 1,
            "会话失效：1s 内单帧 -32603 / 422，不能再发，只能新建", 1_788_790_920_000L, null),
        new MockScenario(CONCURRENT, "[MOCK] 并发被拒 · 单帧",
            "error-concurrent-rejected.jsonl", "load-clean.jsonl", null, 1,
            "同一会话同时只能跑一轮，第二次请求被服务端直接拒绝", 1_788_276_326_000L, null),
        new MockScenario(ACK, "[MOCK] prompt 不派发 · 零帧 · 只有 POP 回执",
            null, null, "0dd3b146c75bf132a65efa7a3080e7cd", 0,
            "prompt 根本没派发：只有 POP 回执、零帧；与断流不同，任务没在跑", 1_789_483_174_000L, null),
        new MockScenario(RENDER, "[MOCK] 渲染覆盖 · 代码块 + mermaid（合成）",
            "synthetic-render.jsonl", "synthetic-render.jsonl", null, 4,
            "合成样例：覆盖代码块高亮与 mermaid 图渲染（非录制件）", 1_789_500_000_000L, null)
    );

    /** 两条"别的来源"的会话，用于证明 SessionSource 过滤真的生效。 */
    private static final String OTHER_SOURCE = "recorded-somewhere-else";
    private static final List<MockScenario> OTHER_SOURCE_SCENARIOS = List.of(
        new MockScenario("mock-other-source-1", "[MOCK] 别的来源 · 应被过滤掉",
            "prompt-short.jsonl", null, null, 22, "这条不该出现在列表里。", 1_788_313_631_000L, OTHER_SOURCE),
        new MockScenario("mock-other-source-2", "[MOCK] 别的来源 · 也应被过滤掉",
            "prompt-short.jsonl", null, null, 22, "这条也不该出现在列表里。", 1_788_313_630_000L, OTHER_SOURCE)
    );

    /** MOCK 模式下"新建会话"登记在这里（进程级，重启即清空）。 */
    private static final CopyOnWriteArrayList<MockScenario> CREATED = new CopyOnWriteArrayList<>();

    public static MockScenario findScenario(String sessionId) {
        for (MockScenario s : CREATED) {
            if (s.sessionId().equals(sessionId)) return s;
        }
        for (MockScenario s : SCENARIOS) {
            if (s.sessionId().equals(sessionId)) return s;
        }
        return null;
    }

    /**
     * MOCK 模式的会话列表。上游字段一律照真实形状给（status 恒 RELEASED、
     * updatedAt===createdAt），因为这两个"没用的字段"本身就是要教的内容：
     * 运行态问不出来，只能靠前端流。
     */
    public static Map<String, Object> mockSessions(String sessionSource) {
        List<MockScenario> all = new ArrayList<>();
        all.addAll(CREATED);
        all.addAll(SCENARIOS);
        all.addAll(OTHER_SOURCE_SCENARIOS);

        List<Map<String, Object>> sessions = new ArrayList<>();
        for (MockScenario s : all) {
            String source = s.sourceOverride() != null ? s.sourceOverride() : sessionSource;
            Map<String, Object> row = new LinkedHashMap<>();
            row.put("sessionId", s.sessionId());
            row.put("title", s.title());
            row.put("createdAt", s.createdAt());
            row.put("updatedAt", s.createdAt());
            row.put("status", "RELEASED");
            row.put("source", source);
            row.put("tags", s.sourceOverride() != null ? List.of() : List.of("mock"));
            row.put("mockScenario", s.teaches());
            sessions.add(row);
        }
        List<Map<String, Object>> kept = new ArrayList<>();
        for (Map<String, Object> s : sessions) {
            if (sessionSource.equals(s.get("source"))) kept.add(s);
        }
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("sessions", kept);
        out.put("filteredOut", sessions.size() - kept.size());
        out.put("total", sessions.size());
        return out;
    }

    /** 新建会话（MOCK）。成功判据与 live 完全一致：只有 SessionId 非空。 */
    public static Map<String, Object> mockCreateSession(String title) {
        String sessionId = SHORT + "-" + Long.toHexString(System.currentTimeMillis());
        CREATED.add(0, new MockScenario(
            sessionId,
            // SessionTitle = 首条 prompt 原文（含注入的校验码说明），与 live 行为一致
            title,
            "prompt-short.jsonl", "load-clean.jsonl", null, 22,
            "MOCK 下新建的会话，回放短轮样例（22 帧 / end_turn）。",
            System.currentTimeMillis(), null));
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("sessionId", sessionId);
        return out;
    }
}

File: server-python/src/das_python/mock_fixtures.py
"""MOCK 模式的数据源：`server-node/test/fixtures/` 下的样例帧流。

单测读的是同一批文件，所以"mock 下能跑通"与"单测绿"指向同一份数据，
不会出现两套真相。与 Node 实现的 server-node/mock/fixtures.ts 同源同语义。
"""

from __future__ import annotations

import json
import time
from pathlib import Path
from typing import Any

from .constants import DEFAULT_SESSION_SOURCE
from .frames import parse_recorded_line

# fixtures 在 Node server 的测试目录里（同一仓库、同一份数据，不复制第二份）。
FIXTURE_DIR = Path(__file__).resolve().parents[3] / "server-node" / "test" / "fixtures"


def read_fixture_frames(name: str) -> list[dict[str, Any]]:
    raw = (FIXTURE_DIR / name).read_text(encoding="utf8")
    frames: list[dict[str, Any]] = []
    for line in raw.split("\n"):
        if not line.strip():
            continue
        frame = parse_recorded_line(line)
        if frame is None:
            raise ValueError(f"fixture {name} 有一行解析不出帧")
        frames.append(frame)
    return frames


def read_fixture_result(name: str) -> dict[str, Any]:
    """非流式录制件：形如 {JsonRpcResponse:{Result:{…}}, RequestId:'…'}。"""
    envelope = read_fixture_envelope(name)
    if envelope["result"] is None:
        raise ValueError(f"fixture {name} 里没有 JsonRpcResponse.Result")
    return envelope["result"]


def read_fixture_envelope(name: str) -> dict[str, Any]:
    parsed = json.loads((FIXTURE_DIR / name).read_text(encoding="utf8"))
    rpc = parsed.get("JsonRpcResponse") or {}
    return {
        "result": rpc.get("Result"),
        "requestId": parsed.get("RequestId"),
        "error": rpc.get("Error"),
    }


class MockScenario:
    def __init__(
        self,
        sessionId: str,  # noqa: N803
        title: str,
        promptFixture: str | None,  # noqa: N803
        historyFixture: str | None,  # noqa: N803
        frameCount: int,  # noqa: N803
        teaches: str,
        createdAt: int,  # noqa: N803
        popAck: str | None = None,  # noqa: N803
        sourceOverride: str | None = None,  # noqa: N815
    ) -> None:
        self.sessionId = sessionId
        self.title = title
        self.promptFixture = promptFixture
        self.historyFixture = historyFixture
        self.popAck = popAck
        self.frameCount = frameCount
        self.teaches = teaches
        self.createdAt = createdAt
        self.sourceOverride = sourceOverride


_SHORT, _TOOLS, _LONG = "mock-short", "mock-tools", "mock-long"
_BREAK, _GHOST, _CONCURRENT, _ACK, _RENDER = (
    "mock-break",
    "mock-ghost",
    "mock-concurrent",
    "mock-ack-only",
    "mock-render",
)

# 八条演示会话：七条各对一份样例帧流（ack-only 那条没有，它演示的正是"零帧"），
# 一条是合成渲染样例（mock-render，非样例录制件）。
MOCK_SCENARIOS: list[MockScenario] = [
    MockScenario(
        sessionId=_SHORT,
        title="[MOCK] 短轮 · end_turn",
        promptFixture="prompt-short.jsonl",
        historyFixture="load-clean.jsonl",
        frameCount=22,
        teaches="最短闭环：一问一答、无工具调用、正常 end_turn",
        createdAt=1_789_026_610_000,
    ),
    MockScenario(
        # marker 注入已退役：标题不再带校验码说明。
        sessionId=_TOOLS,
        title="[MOCK] 工具轮 · 6 次调用",
        promptFixture="prompt-tools.jsonl",
        historyFixture="load-clean.jsonl",
        frameCount=392,
        teaches="工具状态机：6 次调用（5 completed + 1 failed）",
        createdAt=1_789_029_139_000,
    ),
    MockScenario(
        sessionId=_LONG,
        title="[MOCK] 多步分析 · 10 次调用",
        promptFixture="prompt-long.jsonl",
        historyFixture="load-polluted.jsonl",
        frameCount=901,
        teaches="多步分析；历史包含缺少 RequestId 的污染帧",
        createdAt=1_789_027_001_000,
    ),
    MockScenario(
        sessionId=_BREAK,
        title="[MOCK] 断流 · -32603",
        promptFixture="error-stream-break.jsonl",
        historyFixture="load-polluted.jsonl",
        frameCount=1201,
        teaches="SSE 断流：部分回复后收尾 -32603，任务可能仍在服务端跑",
        createdAt=1_788_787_752_000,
    ),
    MockScenario(
        sessionId=_GHOST,
        title="[MOCK] 会话幽灵化 · 422 · 单帧",
        promptFixture="error-session-ghost.jsonl",
        historyFixture=None,
        frameCount=1,
        teaches="会话失效：1s 内单帧 -32603 / 422，不能再发，只能新建",
        createdAt=1_788_790_920_000,
    ),
    MockScenario(
        sessionId=_CONCURRENT,
        title="[MOCK] 并发被拒 · 单帧",
        promptFixture="error-concurrent-rejected.jsonl",
        historyFixture="load-clean.jsonl",
        frameCount=1,
        teaches="同一会话同时只能跑一轮，第二次请求被服务端直接拒绝",
        createdAt=1_788_276_326_000,
    ),
    MockScenario(
        sessionId=_ACK,
        title="[MOCK] prompt 不派发 · 零帧 · 只有 POP 回执",
        promptFixture=None,  # 这一轮一个 ACP 帧都没有，上游只回了一个 POP 层回执就关流
        popAck="0dd3b146c75bf132a65efa7a3080e7cd",
        historyFixture=None,
        frameCount=0,
        teaches="prompt 根本没派发：只有 POP 回执、零帧；与断流不同，任务没在跑",
        createdAt=1_789_483_174_000,
    ),
    MockScenario(
        sessionId=_RENDER,
        title="[MOCK] 渲染覆盖 · 代码块 + mermaid（合成）",
        promptFixture="synthetic-render.jsonl",
        historyFixture="synthetic-render.jsonl",
        frameCount=4,
        teaches="合成样例：覆盖代码块高亮与 mermaid 图渲染（非样例录制件）",
        createdAt=1_789_500_000_000,
    ),
    # 人卡交互（合成）：权限/ask_user_question 弹卡与回覆。
    MockScenario(
        sessionId="mock-permission",
        title="[MOCK] 人卡交互 · 弹卡 + 回覆",
        promptFixture="prompt-permission.jsonl",
        historyFixture="load-clean.jsonl",
        frameCount=3,
        teaches="权限/ask_user_question 弹卡出现，回覆后卡片消失",
        createdAt=1_789_600_000_000,
    ),
]

# 两条"别的来源"的会话，用于证明 SessionSource 过滤真的生效。
_OTHER_SOURCE = "recorded-somewhere-else"
_OTHER_SOURCE_SCENARIOS = [
    MockScenario(
        sessionId="mock-other-source-1",
        title="[MOCK] 别的来源 · 应被过滤掉",
        promptFixture="prompt-short.jsonl",
        historyFixture=None,
        frameCount=22,
        teaches="这条不该出现在列表里。",
        createdAt=1_788_313_631_000,
        sourceOverride=_OTHER_SOURCE,
    ),
    MockScenario(
        sessionId="mock-other-source-2",
        title="[MOCK] 别的来源 · 也应被过滤掉",
        promptFixture="prompt-short.jsonl",
        historyFixture=None,
        frameCount=22,
        teaches="这条也不该出现在列表里。",
        createdAt=1_788_313_630_000,
        sourceOverride=_OTHER_SOURCE,
    ),
]

# MOCK 模式下"新建会话"登记在这里（进程级，重启即清空）。
_created_scenarios: list[MockScenario] = []


def find_scenario(session_id: str) -> MockScenario | None:
    for s in [*_created_scenarios, *MOCK_SCENARIOS]:
        if s.sessionId == session_id:
            return s
    return None


def mock_sessions(session_source: str) -> dict[str, Any]:
    """MOCK 模式的会话列表。

    上游字段一律照真实形状给（SessionStatus 恒 RELEASED、UpdatedAt===CreatedAt），
    因为这两个"没用的字段"本身就是要教的内容：运行态问不出来，只能靠前端流。
    """
    all_scenarios = [*_created_scenarios, *MOCK_SCENARIOS, *_OTHER_SOURCE_SCENARIOS]
    sessions = []
    for s in all_scenarios:
        source = s.sourceOverride or session_source
        sessions.append(
            {
                "sessionId": s.sessionId,
                "title": s.title,
                "createdAt": s.createdAt,
                "updatedAt": s.createdAt,
                "status": "RELEASED",
                "source": source,
                "tags": [] if s.sourceOverride else ["mock"],
                "mockScenario": s.teaches,
            }
        )
    kept = [s for s in sessions if s["source"] == session_source]
    return {
        "sessions": kept,
        "filteredOut": len(sessions) - len(kept),
        "total": len(sessions),
    }


def mock_create_session(title: str) -> dict[str, str]:
    """新建会话（MOCK）。成功判据与 live 完全一致：只有 SessionId 非空。"""
    session_id = f"{_SHORT}-{format(int(time.time() * 1000), 'x')}"
    _created_scenarios.insert(
        0,
        MockScenario(
            sessionId=session_id,
            # SessionTitle = 首条 prompt 原文（含注入的校验码说明），与 live 行为一致
            title=title,
            promptFixture="prompt-short.jsonl",
            historyFixture="load-clean.jsonl",
            frameCount=22,
            teaches="MOCK 下新建的会话，回放短轮样例（end_turn）。",
            createdAt=int(time.time() * 1000),
        ),
    )
    return {"sessionId": session_id}
